### PR TITLE
Refactor to support better handling of data object validation

### LINKED ISSUE #155: Fix uncaught exception error when validating an array object
**Describe the bug**
validate raises an exception while validating a product. 
As the product is invalid, I expect a message indicating what the problem is instead of one indicating there has been an uncaught exception. 

The problem with the product is there is an inconsistency between the number of axes declared and the number of axes defined for an array object. 

    <Array>
      <name>Integration_Time</name>
      <local_identifier>Integration_Time</local_identifier>
      <offset unit="byte">13484630</offset>
      <axes>2</axes>
      <axis_index_order>Last Index Fastest</axis_index_order>
      <description>Integration_Time</description>
      <Element_Array>
        <data_type>SignedLSB8</data_type>
      </Element_Array>
      <Axis_Array>
        <axis_name>index</axis_name>
        <elements>1</elements>
        <sequence_number>1</sequence_number>
      </Axis_Array>
    </Array>

The declared value for Array>axes is 2, but then only one Array>Axis_Array entity has been defined.Changing the axes value from 2 to 1 the exception messages disappear. 

**To Reproduce**
@msbentley will send / attach the product. 

**Expected behavior**
As the product is invalid, a proper message indicating what the problem is instead of one indicating there has been an uncaught exception.

**Version of Software Used**
PDS Validate Tool Report

Configuration:
   Version                       1.17.5
   Date                          2019-11-19T10:01:35Z
...
Product Level Validation Results

  PASS: gov.nasa.pds.validate.ValidateLauncher
      WARNING  [warning.product_not_registered]   Non-registered context products should only be used during archive development. All context products must be registered for a valid, released archive bundle. 

  FAIL: file:/home/fvallejo/TestData/strofio/srn_raw_strofio_re-hq-tof_20190405_20190406.xml
      ERROR  [error.validation.internal_error]   Uncaught exception while validating: null

  FAIL: file:/home/fvallejo/TestData/strofio/srn_raw_strofio_re-lq-tof_20190405_20190406.xml
      ERROR  [error.validation.internal_error]   Uncaught exception while validating: null

Validation complete.

**Desktop (please complete the following information):**
 - OS: Ubuntu 18.04 LTS / Oracle Java 1.8.0_201

### LINKED ISSUE #425: Refactor content validation to more robustly handle intermingled Headers
## 💪 Motivation
...so that we can ensure files with Header / Table / Header / Table will validate successfully (for example).

## 📖 Additional Details

* Follow on to https://github.com/NASA-PDS/validate/issues/416.
* and discussion that started here: https://github.com/NASA-PDS/validate/pull/420#discussion_r734125186

We need to refactor the TableDataContentValidationRule class (and possibly ArrayDataContentValidationRule class as well) per the following identified issues:
1. the logic appears to be assuming that all Headers objects will come before Table objects (and I'm sure Arrays for that matter), but that is not true. They can be intermingled throughout `<File_Area_Observational>` (after `<File>`). See the [PDS4 IM here spec](https://pds.nasa.gov/datastandards/documents/im/v1/index_1G00.html#class_pds_file_area_observational) for more details.
2. solution from https://github.com/NASA-PDS/validate/pull/420 reads the XML label like it is a text file, which seems very dangerous to me (as became apparent in needing to pull out comments from the file). Wherever possible, we should parse XML as an object into Java for more robust access.

## Proposed Solution
Refactor the content validation classes and pds4 jparser to "walk" through all the objects within `<File_Area_Observational>`, and based on the identified object, perform the necessary validation. e.g.

Given a `<File_Area_Observational>` like:

```
<File>
...
</File>
<Header>
</Header>
<Table_Character>
</Table_Character>
<Header>
</Header>
<Table_Character>
</Table_Character>
```

The software would do something like:
```
for obj in fileAreaObservationalObjects
   if obj is Header
      checkHeader
  elif obj is Table_Character
      checkTableCharacter
  elif obj is Table_Binary
      checkTableBinary
...
```

## ⚖️ Acceptance Criteria
**Given** a label with intermingled objects as noted above, but with invalid offsets
**When I perform** validation with content validation enabled
**Then I expect** validate to throw an error about invalid offsets

**Given** a label with intermingled objects as noted above, and valid
**When I perform** validation with content validation enabled
**Then I expect** validate to complete successfully with no errors

## Test Data
* See https://github.com/NASA-PDS/validate/issues/361#issuecomment-951643479 for array test data
* See https://github.com/NASA-PDS/validate/issues/416 for table

### LINKED ISSUE #453: Validate should not check PDF/A validity if content validation is disabled
## 🐛 Describe the bug

When I execute validate with `--skip-content-validation` flag, it still throws an invalid PDF/A warning:

```
  PASS: file:/data/pds4/test-data/registry/lab.hydrocarbon_spectraV1/document/h2205k292k.xml
      WARNING  [warning.pdf.file.not_pdfa_compliant]   file:/data/pds4/test-data/registry/lab.hydrocarbon_spectraV1/document/h2205k292k.pdf is not valid PDF/A file or does not exist
        56 product validation(s) completed
```

## 📜 To Reproduce
```
validate -R pds4.bundle $TEST_DATA_HOME/registry/lab.hydrocarbon_spectraV1/bundle.lab.hydrocarbon_spectra.xml --skip-content-validation
```

## 🕵️ Expected behavior

Validate would skip trying to check validity of the PDF because we are disabling content validation.

## 📚 Version of Software Used

2.0.7, 2.2.0-SNAPSHOT

## 🩺 Test Data / Additional context

`pds-dev3:$TEST_DATA_HOME/registry/lab.hydrocarbon_spectraV1/`

---

### LINKED ISSUE #473: NullPointerException when Table_Delimited is missing records attribute
## 🐛 Describe the bug

NullPointerException thrown on invalid label. The software should fail gracefully with an error, not throw an uncaught exception. Reported by B. Deen.

## 📜 To Reproduce
```
% validate-2.1.4/bin//validate -target !* --report-file !*.valrpt --schematron /project/m20/rgd/pds4/validate/dd//*.sch --schema /project/m20/rgd/pds4/validate/dd//*.xsd

% valgo mech_drive.csv.xml
java.lang.reflect.InvocationTargetException
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at gov.nasa.pds.tools.validate.rule.AbstractValidationRule.execute(AbstractValidationRule.java:71)
	at org.apache.commons.chain.impl.ChainBase.execute(ChainBase.java:191)
	at gov.nasa.pds.tools.validate.task.ValidationTask.execute(ValidationTask.java:133)
	at gov.nasa.pds.tools.validate.task.BlockingTaskManager.submit(BlockingTaskManager.java:27)
	at gov.nasa.pds.tools.label.LocationValidator.validate(LocationValidator.java:275)
	at gov.nasa.pds.validate.ValidateLauncher.doValidation(ValidateLauncher.java:1333)
	at gov.nasa.pds.validate.ValidateLauncher.processMain(ValidateLauncher.java:1590)
	at gov.nasa.pds.validate.ValidateLauncher.main(ValidateLauncher.java:1636)
Caused by: java.lang.NullPointerException
	at gov.nasa.pds.tools.validate.rule.pds4.TableDataContentValidationRule.validateTableDataContents(TableDataContentValidationRule.java:951)
	... 12 more
```

## 🕵️ Expected behavior

Expected validation report:
```
PDS Validate Tool Report

Configuration:
   Version                       2.1.5-SNAPSHOT
   Date                          2022-01-28T00:42:48Z

Parameters:
   Targets                       [file:/Users/jpadams/Downloads/mech_drive.csv.xml]
   Severity Level                WARNING
   Recurse Directories           true
   File Filters Used             [*.xml, *.XML]
   Data Content Validation       off
   Product Level Validation      on
   Max Errors                    100000
   Registered Contexts File      /Users/jpadams/Documents/proj/pds/pdsen/workspace/validate/validate-2.1.5-SNAPSHOT/resources/registered_context_products.json



Product Level Validation Results

  FAIL: file:/Users/jpadams/Downloads/mech_drive.csv.xml
      ERROR  [error.label.schema]   line 18, 17: cvc-pattern-valid: Value '3.0.0' is not facet-valid with respect to pattern '([0-9]+)(\.){1}([0-9]+)' for type 'version_id'.
      ERROR  [error.label.schema]   line 18, 17: cvc-type.3.1.3: The value '3.0.0' of element 'version_id' is not valid.
      ERROR  [error.label.schema]   line 94, 25: cvc-complex-type.2.4.a: Invalid content was found starting with element '{"http://pds.nasa.gov/pds4/pds/v1":record_delimiter}'. One of '{"http://pds.nasa.gov/pds4/pds/v1":records}' is expected.
      ERROR  [error.label.schema]   line 96, 25: cvc-complex-type.2.4.b: The content of element 'Record_Delimited' is not complete. One of '{"http://pds.nasa.gov/pds4/pds/v1":maximum_record_length, "http://pds.nasa.gov/pds4/pds/v1":Field_Delimited, "http://pds.nasa.gov/pds4/pds/v1":Group_Field_Delimited}' is expected.
      ERROR  [error.label.schematron]   line 96, 25: The values for 'fields' and 'groups' must not both be '0'.  The number of fields found: (0).  The number of groups found: (0).
        1 product validation(s) completed

Summary:

  5 error(s)
  0 warning(s)

  Product Validation Summary:
    0          product(s) passed
    1          product(s) failed
    0          product(s) skipped

  Referential Integrity Check Summary:
    0          check(s) passed
    0          check(s) failed
    0          check(s) skipped

  Message Types:
    4            error.label.schema
    1            error.label.schematron

End of Report
Completed execution in 7720 ms
```

## 📚 Version of Software Used

2.1.4

## 🩺 Test Data / Additional context

[mech_drive.csv.txt](https://github.com/NASA-PDS/validate/files/7954872/mech_drive.csv.txt)
[mech_drive.csv.xml.txt](https://github.com/NASA-PDS/validate/files/7954873/mech_drive.csv.xml.txt)

### LINKED ISSUE #480: Validate does not calculate overlaps correctly when Header is not first object in file
## 🐛 Describe the bug
In a file with Table_Character1 + Header + Table_Character2, Validate reports that Table_Character1 cannot have offset 0 because there is already an object with length equal to Header in front of it.

## 📜 To Reproduce
1. Unzip and untar attached file header_problem.tar.gz
2. Run Validate on test.xml
3. See error:
 FAIL: file:/Users/rsimpson/Data/DAWN/dawn-rss-raw-vesta/calib-wea/wea/test.xml
    Begin Content Validation: file:/Users/rsimpson/Data/DAWN/dawn-rss-raw-vesta/calib-wea/wea/test.tab
      ERROR  [error.table.fields_mismatch]   table 1: The table offset 0 for object 'Identifier Record for Day-of-Year 001 (Block 1)' is invalid.  The previously defined object ends at byte 180
    End Content Validation: file:/Users/rsimpson/Data/DAWN/dawn-rss-raw-vesta/calib-wea/wea/test.tab
        1 product validation(s) completed

Commenting out the Header definition in the label makes the error go away.  Changing  Header.object_length  to  'nnn'  changes the error message to "The previously defined object ends at byte nnn".
Error does not seem to depend on IM.  I am trying to validate data using IM 1.13.0.0; but I also got errors when I changed to IM 1.16.0.0.

[header_problem.tar.gz](https://github.com/NASA-PDS/validate/files/8150712/header_problem.tar.gz)

## 🕵️ Expected behavior
No error should be reported

## 📚 Version of Software Used
validate 2.1.4

## 🩺 Test Data / Additional context
See attached header_problem.tar.gz

## 🖥 System Info
 - Mac OS 12.0.1
 - Safari 15.1 (17612.2.9.1.20)

---

### LINKED ISSUE #481: B13.1 Content Validation Improvements: Intermingled Headers
## 💡 Description

See sub-tasks

### LINKED ISSUE #537: Refactor `getMessageCountBasedOnProblemType` function using introspection
This `getMessageCountBasedOnProblemType`function omg my eyes 👀 😆

Maybe use introspection here? The reflection API could turn this 113 line function into 8–10 lines 🤔

_Originally posted by @nutjob4life in https://github.com/NASA-PDS/validate/pull/536#discussion_r955411163_

### PR DESCRIPTION
**Pull request by jordanpadams**
## 🗒️ Summary
Refactor to support better handling of data object validation and intermingled headers/tables/arrays

Highlights include
    * combine all data object (table and array) validation to read from top to bottom, instead of individually
    * utilize new PDS4 Jparser objects to track data object location
    * fix issues with checking data definitions in terms of offsets vs. expected object lengths (#480, #425)
    * no longer reads PDFs when content validation is disabled (#453)
    * improves error message counting in `getMessageCountBasedOnProblemType` (#537)

Tightly coupled with NASA-PDS/pds4-jparser@f31281fbf0059714c12235ec719b896b28108d9a

Requires merge of https://github.com/NASA-PDS/pds4-jparser/pull/60

## ⚙️ Test Data and/or Report
See Github Actions

## ♻️ Related Issues

Resolves #480 
Resolves #425 
Resolves #481 
Resolves #453
Resolves #537 
Resolves #473
resolves #155

### CODE REVIEW

File: src/main/java/gov/nasa/pds/tools/util/PDFUtil.java [hunk 1]
```diff
@@ -99,11 +99,11 @@ private synchronized void writeErrorToFile(URI uri, ValidationResult result, Str
     }
   }
 
-  private boolean validatePDF(URI uri, String pdfRef) {
+  private boolean validatePDF(URI uri, String pdfRef) throws IOException {
     boolean pdfValidateFlag = false;
 
-    // Create a parser and auto-detect flavour
     try {
+      // Create a parser and auto-detect flavour
```
Comment (nutjob4life) [inline src/main/java/gov/nasa/pds/tools/util/PDFUtil.java:106]: 🇬🇧

File: src/main/java/gov/nasa/pds/tools/validate/content/array/ArrayContentValidator.java [hunk 2]
```diff
@@ -120,11 +121,6 @@ public void validate(Array array, ArrayObject arrayObject) {
     try {
       process(array, arrayObject, dimensions, new int[dimensions.length], 0, dimensions.length - 1);
 
-    } catch (IOException io) {
-      listener.addProblem(new ArrayContentProblem(
-          new ProblemDefinition(ExceptionType.FATAL, ProblemType.ARRAY_DATA_FILE_READ_ERROR,
-              "Error occurred while reading data file: " + io.getMessage()),
-          dataFile, label, arrayIndex, null));
     } catch (Exception e) {
```
Comment (nutjob4life) [inline src/main/java/gov/nasa/pds/tools/validate/content/array/ArrayContentValidator.java:124]: Catching general `Exception` on line 128 also catches `RuntimeException` due to the [broken exception hierarchy in Java](https://github.com/NASA-PDS/nasa-pds.github.io/wiki/Java-Exceptions-Gotcha).

If we can't catch a more specific exception here, then `RuntimeException` should be caught first and re-thrown, then the general `Exception`.

File: src/main/java/gov/nasa/pds/tools/validate/content/array/ArrayContentValidator.java [hunk 3]
```diff
@@ -165,6 +162,7 @@ private void process(Array array, ArrayObject arrayObject, int[] dimensions, int
         }
       }
     }
+    arrayObject.close();
```
Comment (nutjob4life) [inline src/main/java/gov/nasa/pds/tools/validate/content/array/ArrayContentValidator.java:165]: Should `arrayObject.close()` occur in a `finally` clause?

File: src/main/java/gov/nasa/pds/tools/validate/rule/pds4/ArrayValidator.java [hunk 4]
```diff
@@ -0,0 +1,226 @@
...
+
+    LOG.debug("validateArray:targetFileName {}", targetFileName);
+
+    try {
+      // Check if the data type is something we support
+      String dataType = "";
+      try {
+        dataType = this.array.getArray().getElementArray().getDataType();
+        Enum.valueOf(NumericDataType.class, dataType);
+      } catch (IllegalArgumentException a) {
```
Comment (nutjob4life) [inline src/main/java/gov/nasa/pds/tools/validate/rule/pds4/ArrayValidator.java:88]: Looks like the exception instance isn't used in the handler? If not, it'd be nice to see a more explanatory name here than `a`, such as `ignored` or `notUsed`, etc.

File: src/main/java/gov/nasa/pds/tools/validate/rule/pds4/ArrayValidator.java [hunk 5]
```diff
@@ -0,0 +1,226 @@
...
+        throw new IllegalArgumentException(dataType + " is not supported at this time.");
+      }
+
+      // Verify that the elements match the object statistics defined
+      // within their associated label, if they exist
+      if (!this.array.getArray().getAxisArraies().isEmpty()) {
```
Comment (nutjob4life) [inline src/main/java/gov/nasa/pds/tools/validate/rule/pds4/ArrayValidator.java:94]: [arraies](https://www.tripadvisor.com/LocationPhotoDirectLink-g150812-d13553868-i324462110-Paz-Playa_del_Carmen_Yucatan_Peninsula.html)?

File: src/main/java/gov/nasa/pds/tools/validate/rule/pds4/DataDefinitionAndContentValidationRule.java [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+        checkOffset(target, minimumExpectedOffset, obj.getOffset(), objectCounter);
+        LOG.debug("class: {}", obj.getClass());
+        LOG.debug("label offset: {}, label size: {}", obj.getOffset(), obj.getSize());
+        LOG.debug("minimumExpectedOffset: {}", minimumExpectedOffset);
+
+        // Set expected offset for next loop through
+        minimumExpectedOffset = obj.getOffset() + obj.getSize();
+
+        // Check and validate per specific object type
+        if (obj instanceof TableObject) {
```
Comment (nutjob4life) [inline src/main/java/gov/nasa/pds/tools/validate/rule/pds4/DataDefinitionAndContentValidationRule.java:81]: If there will be any other kinds of `obj` here, consider replacing `if-else` with a virtual constructor to be more "object oriented" 😄

I could even see the superclass of both `TableObject` and `ArrayObject` have a virtual method to `get my own validator` which the subclasses override to return `TableValidator`, `ArrayValidator`, etc.

File: src/main/java/gov/nasa/pds/tools/validate/rule/pds4/DataDefinitionAndContentValidationRule.java [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+    } catch (InvalidTableException e) {
+      String msg = "Data Object #" + objectCounter + ": " + e.getMessage();
+      getListener().addProblem(new ValidationProblem(
+          new ProblemDefinition(ExceptionType.ERROR, ProblemType.TABLE_DEFINITION_PROBLEM, msg),
+          getTarget(), objectCounter, -1));
+    } catch (ParseException e) {
+      // Ignore. Shouldn't happen
+    } catch (FileNotFoundException e) {
+      // File not found. This is caught and error is tracked elsewhere
+    } catch (Exception e) {
```
Comment (nutjob4life) [inline src/main/java/gov/nasa/pds/tools/validate/rule/pds4/DataDefinitionAndContentValidationRule.java:111]: Same `RuntimeException`/`Exception` gotcha

**Review by nutjob4life — APPROVED** [review]
Some minor nits only—otherwise looks great

File: src/main/java/gov/nasa/pds/tools/util/PDFUtil.java (hunk 1 above)
Comment (jordanpadams) [reply src/main/java/gov/nasa/pds/tools/util/PDFUtil.java:106]: @nutjob4life 😆 yeah. That is how they spell in the verapdf library so we follow them!

File: src/main/java/gov/nasa/pds/tools/validate/content/array/ArrayContentValidator.java (hunk 3 above)
Comment (jordanpadams) [reply src/main/java/gov/nasa/pds/tools/validate/content/array/ArrayContentValidator.java:165]: Definitely

File: src/main/java/gov/nasa/pds/tools/validate/rule/pds4/ArrayValidator.java (hunk 5 above)
Comment (jordanpadams) [reply src/main/java/gov/nasa/pds/tools/validate/rule/pds4/ArrayValidator.java:94]: @nutjob4life yeah... I think that is actually a JAXB thing for whatever reason?

File: src/main/java/gov/nasa/pds/tools/validate/content/array/ArrayContentValidator.java (hunk 2 above)
Comment (jordanpadams) [reply src/main/java/gov/nasa/pds/tools/validate/content/array/ArrayContentValidator.java:124]: 👍 similar to jparser, this code is LITTERED with overuse of `Exception`. Should have a ticket to scrub the whole codebase for this. I actually wish sonatype yelled (via GitHub comments) at people for this

File: src/main/java/gov/nasa/pds/tools/validate/rule/pds4/ArrayValidator.java (hunk 4 above)
Comment (jordanpadams) [reply src/main/java/gov/nasa/pds/tools/validate/rule/pds4/ArrayValidator.java:88]: 👍

File: src/main/java/gov/nasa/pds/tools/validate/rule/pds4/DataDefinitionAndContentValidationRule.java (hunk 6 above)
Comment (jordanpadams) [reply src/main/java/gov/nasa/pds/tools/validate/rule/pds4/DataDefinitionAndContentValidationRule.java:81]: 👍

Comment (jordanpadams) [reply src/main/java/gov/nasa/pds/tools/validate/rule/pds4/DataDefinitionAndContentValidationRule.java:81]: We ready do this with TableAdapter, just need to augment something similar for this (or even merge all of this *Adapter and *Object stuff together)

File: src/main/java/gov/nasa/pds/tools/validate/rule/pds4/DataDefinitionAndContentValidationRule.java (hunk 7 above)
Comment (jordanpadams) [reply src/main/java/gov/nasa/pds/tools/validate/rule/pds4/DataDefinitionAndContentValidationRule.java:111]: 👍

**Pull request merged**